src/main/java/com/linecorp/armeria/common/stream/PublisherBasedStreamMessage.java

```diff
@@ -0,0 +1,264 @@
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        subscribe0(subscriber, executor);
+    }
+
+    private void subscribe0(Subscriber<? super T> subscriber, Executor executor) {
+        final SubscriberWrapper s = new SubscriberWrapper(this, subscriber, executor);
+        if (!subscriberUpdater.compareAndSet(this, null, s)) {
+            if (this.subscriber == ABORTED_SUBSCRIBER) {
+                throw new IllegalStateException("cannot subscribe to an aborted publisher");
+            } else {
+                throw new IllegalStateException(
+                        "subscribed by other subscriber already: " + subscriber);
+            }
+        }
+
+        publisher.subscribe(s);
+    }
+
+    @Override
+    public void abort() {
+        final AbortableSubscriber s = subscriberUpdater.getAndSet(this, ABORTED_SUBSCRIBER);
+        if (s != null) {
+            s.abort();
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> closeFuture() {
+        return closeFuture;
+    }
+
+    private interface AbortableSubscriber extends Subscriber<Object> {
+        void abort();
+    }
+
+    private static final class AbortedSubscriber implements AbortableSubscriber {
+        @Override
+        public void abort() {}
+
+        @Override
+        public void onSubscribe(Subscription s) {}
+
+        @Override
+        public void onNext(Object o) {}
+
+        @Override
+        public void onError(Throwable t) {}
+
+        @Override
+        public void onComplete() {}
+    }
+
+    @VisibleForTesting
+    static final class SubscriberWrapper implements AbortableSubscriber {
+        private final PublisherBasedStreamMessage<?> parent;
+        private final Subscriber<Object> subscriber;
+        private final Executor executor;
+        private volatile boolean abortPending;
+        private volatile SubscriptionWrapper subscription;
+
+        @SuppressWarnings("unchecked")
+        SubscriberWrapper(PublisherBasedStreamMessage<?> parent,
+                          Subscriber<?> subscriber, Executor executor) {
+            this.parent = parent;
+            this.subscriber = (Subscriber<Object>) subscriber;
+            this.executor = executor;
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            final Executor executor = this.executor;
+            final SubscriptionWrapper wrappedSubscription = new SubscriptionWrapper(parent, s);
+            subscription = wrappedSubscription;
+            if (executor == null) {
+                onSubscribe0(wrappedSubscription);
+            } else {
+                executor.execute(() -> onSubscribe0(wrappedSubscription));
+            }
+        }
+
+        private void onSubscribe0(Subscription s) {
+            try {
+                subscriber.onSubscribe(s);
+            } finally {
+                if (abortPending) {
+                    subscription.cancel();
```

Comment:
I mistakenly thought `s` is not `SubscriptionWrapper`. It should be `s.cancel()`. Sorry about confusion!